xtask/src/main.rs

```diff
@@ -0,0 +1,605 @@
+    let mut out = String::new();
+    stdout
+        .read_to_string(&mut out)
+        .with_context(|| "could not read standard output")?;
+    Ok(out)
+}
+
+fn do_test<F>(t: F, context: &str, errors: &mut Vec<String>)
+where
+    F: FnOnce() -> Result<()>,
+{
+    match t() {
+        Ok(_) => {}
+        Err(e) => errors.push(format!("{}: {}", context, e)),
+    }
+}
+
+fn rustc_is_nightly() -> bool {
+    // if this crashes the system is not in a good state, so we'll not pretend to be able to recover
+    let out = run_capturing_stdout(Command::new("rustc").args(&["-V"])).unwrap();
+    out.contains("nightly")
+}
+
+fn load_expected_output(name: &str, release_mode: bool) -> Result<String> {
+    let path = Path::new("firmware/qemu/src/bin");
+
+    let filename;
+    if release_mode {
+        filename = format!("{}.release.out", name);
+    } else {
+        filename = format!("{}.out", name);
+    }
+
+    let path = path.join(filename);
+
+    // for error context closure
+    let path_str = path.to_str().unwrap_or("(non-Unicode path)").to_string();
+
+    let content = fs::read_to_string(path)
+        .with_context(|| format!("Failed to load expected output data from {}", path_str))?;
+    Ok(content)
+}
+
+fn test_single_snapshot(name: &str, features: &str, release_mode: bool) -> Result<()> {
+    let display_name = format!(
+        "{} ({})",
+        name,
+        if release_mode { "release" } else { "dev" }
+    );
+    println!("{}", display_name);
+    let cwd_name = "firmware/qemu".to_string();
+    let mut args;
+    if release_mode {
+        args = vec!["-q", "rrb", name]
+    } else {
+        args = vec!["-q", "rb", name]
+    }
+    if !features.is_empty() {
+        args.extend_from_slice(&["--features", features]);
+    }
+
+    let actual = run_capturing_stdout(Command::new("cargo").args(&args).current_dir(cwd_name))?;
+
+    let expected = load_expected_output(name, release_mode)?;
+
+    let diff = TextDiff::from_lines(&expected, &actual);
+
+    // if anything isn't ChangeTag::Equal, print it and turn on error flag
+    let mut actual_matches_expected = true;
+    for op in diff.ops() {
+        for change in diff.iter_changes(op) {
+            let styled_change = match change.tag() {
+                ChangeTag::Delete => Some(("-", Style::new().red())),
+                ChangeTag::Insert => Some(("+", Style::new().green())),
+                ChangeTag::Equal => None,
+            };
+            if let Some((sign, style)) = styled_change {
+                actual_matches_expected = false;
+                eprint!("{}{}", style.apply_to(sign).bold(), style.apply_to(change),);
+            }
+        }
+    }
+
+    if actual_matches_expected {
+        Ok(())
+    } else {
+        Err(anyhow!("{}", display_name))
+    }
+}
+
+fn get_installed_targets() -> Result<HashSet<String>> {
+    let out = run_capturing_stdout(Command::new("rustup").args(&["target", "list"]))?;
+    let installed_marker = " (installed)";
```

Comment:
make `installed_marker` a `const`?

Reply:
I'm not A BIG FAN OF CAPS IN CODE but I guess that can be done

Reply:
actually `let` is also fine in rust, since it is immutable, but somehow `const` still communicates this better for me 🤷🏾‍♀️

Reply:
`const` *does* indicate that it's fixed at compile time, so yeah